### PR TITLE
aa/attester: add IBM Secure Execution driver framework

### PR DESCRIPTION
**Pull request by huoqifeng**
This is the initial draft to fix: https://github.com/confidential-containers/guest-components/issues/485 and the draft PR provides the API framework without a real pvattest crate yet.

Try it after adding the new Tee type `Tee::Se` in https://github.com/virtee/kbs-types/blob/main/src/lib.rs#L24, run:
```
cd guest-components
TEE_PLATFORM=se KBC=cc_kbc make
```
Depends on: 
- [x] https://github.com/virtee/kbs-types/issues/26

### CODE REVIEW

**Comment by Xynnn007** [comment]
Thanks @huoqifeng for the contribution! I'd love to review this if it is ok : D

**Comment by huoqifeng** [comment]
@Xynnn007 this is not fully ready for review, but I think we can start discuss the potential changes both on guest-components side and on trustee side, which is tracked by https://github.com/confidential-containers/trustee/pull/345.

We want to move the `challenge` generation from session to attestation-service and introduce a new fn `generate_challenge`. like in https://github.com/huoqifeng/trustee/blob/s390x-verifier/attestation-service/attestation-service/src/lib.rs#L243, we want use this `challenge.extra_params` in guest like https://github.com/huoqifeng/guest-components/blob/s390x-attester/attestation-agent/kbs_protocol/src/client/rcar_client.rs#L140

The reason is: for IBM Secure Execution (SE), `attester` need an `attestation-request` to generate the `evidence`, while the `attestation-request` should always be created on `verifier`. CC @BbolroC  @steffen-eiden

**Comment by huoqifeng** [comment]
@Xynnn007 @fitzthum I'd like start the review process of the API change to support IBM SE. May you please help on that?

File: attestation-agent/attester/src/se/seattest.rs
Comment (fitzthum) [inline attestation-agent/attester/src/se/seattest.rs:1]: Might want to rename this file. It looks like it says "seat test"

File: attestation-agent/attester/src/se/seattest.rs
Comment (fitzthum) [inline attestation-agent/attester/src/se/seattest.rs:13]: This trait doesn't seem necessary to me. Why not just put the logic in the methods above (and call into the real SE crate from there once it is ready)? Also, `SeImplAttester` is a slightly strange name for a trait.

File: attestation-agent/kbs_protocol/test/policy.rego
Comment (fitzthum) [inline attestation-agent/kbs_protocol/test/policy.rego:11]: I think this policy file is intended to be used with the nontee (sample tests) so I wouldn't add the SE clause here. Isn't the fake SE attester always returning false for platform detection anyway?

**Review by fitzthum — COMMENTED** [review]
A few high-level comments. I am still thinking about the bigger changes to the protocol.

File: attestation-agent/attester/src/se/seattest.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/seattest.rs:1]: Thanks @fitzthum , we'll put the crate in https://crates.io/crates/s390_pv, it's still in-progress.

File: attestation-agent/attester/src/se/seattest.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/seattest.rs:13]: Thanks @fitzthum , we'll put the real crate in https://crates.io/crates/s390_pv, it's still in-progress. So, `FakeSeAttest` and `SeImplAttester` will be replaced the crate eventually.

File: attestation-agent/kbs_protocol/test/policy.rego
Comment (huoqifeng) [reply attestation-agent/kbs_protocol/test/policy.rego:11]: Yeah, this is what I added when testing se driver, we need drop it when merge.

File: attestation-agent/kbs_protocol/src/client/rcar_client.rs
Comment (mkulke) [inline attestation-agent/kbs_protocol/src/client/rcar_client.rs:196]: do we want to have this as default behaviour? other TEE  impls might have `extra_params` set and still want to use the hashed runtime_data.

we have the `tee` available on the calling function site, so we can do a special case handling.

Comment (huoqifeng) [reply attestation-agent/kbs_protocol/src/client/rcar_client.rs:196]: Yes, do some special for IBM SE and other TEE based on `tee` is alternative and I think it's hood also.

Comment (huoqifeng) [reply attestation-agent/kbs_protocol/src/client/rcar_client.rs]: Fixed

File: attestation-agent/kbs_protocol/test/policy.rego
Comment (huoqifeng) [reply attestation-agent/kbs_protocol/test/policy.rego]: Fixed.

File: attestation-agent/attester/src/se/seattest.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/seattest.rs:1]: Fixed

File: attestation-agent/attester/src/se/seattest.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/seattest.rs]: Fixed.

**Comment by fitzthum** [comment]
@huoqifeng Are we waiting for the underlying SE crate before we merge this? Any idea when it will be ready?

**Comment by huoqifeng** [comment]
@fitzthum because the code `tools/client` in trustee PR depends on the `kbs_protocol` in guest-components, I'm think we merge this PR after review completed without the underlying IBM SE crate, and then add it after the PR in trustee merged. So the steps will be:
- Review and merge  https://github.com/confidential-containers/guest-components/pull/492 without IBM SE crate
- Review and merge https://github.com/confidential-containers/trustee/pull/345/ with IBM SE crate (@Xynnn007 suggest we should add the real crate in trustee)
- Add IBM SE crate in guest-components in a new PR
@fitzthum @Xynnn007 wdyt?

**Review by Xynnn007 — COMMENTED** [review]
There are some nits. Let's make it better

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs:115]: UserData is optional check point, it's not checked on trustee side yet. We're still in discussion what value should be put in.

File: attestation-agent/attester/src/se/ibmse.rs
Comment (mkulke) [inline attestation-agent/attester/src/se/ibmse.rs:22]: Can we move this to attester/src/lib, since the whole module doesn't make sense if the target arch is not s390x?

File: attestation-agent/attester/Cargo.toml
Comment (huoqifeng) [reply attestation-agent/attester/Cargo.toml]: Fixed, thanks!

File: attestation-agent/attester/Cargo.toml
Comment (huoqifeng) [reply attestation-agent/attester/Cargo.toml]: Fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: Fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: Fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: Fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: Fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: Fixed, thanks!

File: Cargo.toml
Comment (huoqifeng) [reply Cargo.toml]: Fixed, thanks!

File: attestation-agent/kbs_protocol/src/client/rcar_client.rs
Comment (huoqifeng) [reply attestation-agent/kbs_protocol/src/client/rcar_client.rs]: Fixed, thanks!

File: attestation-agent/attester/src/se/mod.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs]: Fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: Fixed, thanks!

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs]: Changed it to optional for now.

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs:18]: @mkulke there is a `#[cfg(feature = "se-attester")]` to build this module selectively, is that OK for you?

Comment (mkulke) [reply attestation-agent/attester/src/se/ibmse.rs:18]: to understand the reasoning behind this additional flag: the motivation is to compile the se-attester on all archs into the bin but disable the code at runtime on the wrong archs?

Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs:18]: Yes, intention is to use `#[cfg(feature = "se-attester")]` to include `se` or not, `#[cfg(not(target_arch = "s390x"))]` to make sure it works well also on x86 if `se-attester` selected.

Comment (mkulke) [reply attestation-agent/attester/src/se/ibmse.rs:18]: But is this what we want? As I understand se-attester would never work on x86_64, why package the code into the bins?

File: attestation-agent/attester/Cargo.toml
Comment (mkulke) [inline attestation-agent/attester/Cargo.toml:18]: is pv_core sufficient maybe, to avoid linking to openssl?

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs:18]: Not sure whether its reasonable, I want it also be checked by CI before we add a s390x runner in CI and GHA.

Comment (mkulke) [reply attestation-agent/attester/src/se/ibmse.rs:18]: ok, to be fair we have the same for the arm attester, we'll include it in the x86_64 bins even if it's dead code. we can leave it like this, I suppose, but maybe check whether we can use pv_core instead to get rid of the openssl dep (it might cause issues in initrd-based guest-components)

Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs:18]: oh, I got your concern!

File: attestation-agent/attester/Cargo.toml
Comment (huoqifeng) [reply attestation-agent/attester/Cargo.toml:18]: Got your real concern.

File: attestation-agent/attester/src/se/ibmse.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/ibmse.rs:18]: Fixed, thanks!

**Comment by huoqifeng** [comment]
@mkulke @Xynnn007 comments addressed, thanks for all of your review! I'm curious why the CI is not health with no failure task?

**Comment by mkulke** [comment]
> @mkulke @Xynnn007 comments addressed, thanks for all of your review! I'm curious why the CI is not health with no failure task?

some workflows have been cancelled, seemingly for no reason. I've noticed this problem elsewhere this week, I'll re-schedule the cancelled ones, that worked so far.

**Review by mkulke — APPROVED** [review]
lgtm

File: attestation-agent/attester/Cargo.toml
Comment (Xynnn007) [inline attestation-agent/attester/Cargo.toml:56]: We'd better add a comment here.
```
`se-attester` feature can work only on `s390x`  target arch.
```

File: attestation-agent/attester/src/se/ibmse.rs
Comment (Xynnn007) [inline attestation-agent/attester/src/se/ibmse.rs:19]: As we already mark the code path here would only be executed when `target_arch` is `s390x` [here](https://github.com/confidential-containers/guest-components/pull/492/files#diff-8fd997b18e87172dbdcec4c6ab0268c8d5b8e4dcd7b3a4d493ab1f9519d6c10aR137-R138), we might not need this condition.

File: attestation-agent/attester/src/se/mod.rs
Comment (Xynnn007) [inline attestation-agent/attester/src/se/mod.rs:36]: I am afraid that this unit test cannot be run. As the input `report_data` will be treated as a `SeAttestationRequest` and the test here defines an empty one. Thus I suggest either
1. add a real worked `report_data` in under `https://github.com/confidential-containers/guest-components/tree/main/attestation-agent/attester/test_data` (requires a new dir creation)
2. delete this unit test

File: attestation-agent/attester/src/se/ibmse.rs
Comment (Xynnn007) [inline attestation-agent/attester/src/se/ibmse.rs:105]: I noticed that the logic under `perform` is direct and compresihensive. Could we move all the logic directly to `get_evidence` in `se/mod.rs`. Also, we might not need extra definition of `calc_userdata()`, could we just 
```rust
let userdata = UserData {
        image_btph: image_btph.into(),
    };
```

**Review by Xynnn007 — COMMENTED** [review]
Some more nits..

**Comment by huoqifeng** [comment]
@Xynnn007 I guess the failure in https://github.com/confidential-containers/guest-components/actions/runs/9358373555/job/25760080271?pr=492 is caused by network connection?

**Comment by huoqifeng** [comment]
@mkulke @Xynnn007 thank you very much for reviewing the PR and gave so many useful suggestions! thanks for your approval.

File: attestation-agent/attester/src/se/mod.rs
Comment (Xynnn007) [inline attestation-agent/attester/src/se/mod.rs:104]: I think we might not need `.clone()` here. Every `.clone()` would alloc a slice of memory.

**Review by Xynnn007 — APPROVED** [review]
Looks good. Thanks @huoqifeng .

File: attestation-agent/attester/src/se/mod.rs
Comment (Xynnn007) [inline attestation-agent/attester/src/se/mod.rs:78]: Also this `.clone()`
```suggestion
        debug!("userdata json: {userdata:#?}");
```

File: attestation-agent/attester/src/se/mod.rs
Comment (Xynnn007) [inline attestation-agent/attester/src/se/mod.rs:84]: I noticed that the `user_data` was to be validated [on the AS side](https://github.com/confidential-containers/trustee/pull/345/files#diff-915457db8ca21ef25fda79ebeadd19a81a62971717017b3d7ffc06b9bc00ebceR197) . Also the [document of attester](https://docs.rs/s390_pv/0.10.0/s390_pv/uv/struct.AttestationCmd.html#method.new_request) and [verifier side](https://docs.rs/s390_pv/0.10.0/s390_pv/attest/struct.AttestationItems.html#method.new) show this nonce is generated before measurement. I wonder if this value should be generated on the AS side first, and then do the check -- like a challenge value to prevent replay attack, also just like `report_data`/`runtime_data` for other platforms.

Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs:84]: user_data is optional data which will be injected on attester side and checked on AS side saved beforehand -- yes, which is optional today. We are going to add more info such as initdata hash in user_data in future. But it's not used today. We check OS Image and Firmware info only today. @Xynnn007

File: attestation-agent/attester/src/se/mod.rs
Comment (fitzthum) [inline attestation-agent/attester/src/se/mod.rs:19]: Why not just return false in the else case here?

And for the next step, can you compare the string directly rather than converting it to an int.

Also if we need the variable v, can we rename it to something more descriptive?

File: attestation-agent/attester/src/se/mod.rs
Comment (fitzthum) [inline attestation-agent/attester/src/se/mod.rs:73]: Can you add a comment about what `image_btph` is? There are a lot of terms here that I am not familiar with. Is there something we could add to the docs to describe the flow in general?

File: attestation-agent/kbs_protocol/src/client/rcar_client.rs
Comment (fitzthum) [inline attestation-agent/kbs_protocol/src/client/rcar_client.rs:191]: not sure we need this

**Review by fitzthum — COMMENTED** [review]
Looks pretty good. A couple minor comments.

File: attestation-agent/attester/src/se/mod.rs
Comment (steffen-eiden) [reply attestation-agent/attester/src/se/mod.rs:19]: Using an int completely makes sense.
He took the code from [pv_guest_bit_set](https://docs.rs/s390_pv/0.10.0/s390_pv/misc/fn.pv_guest_bit_set.html) in the `s390_pv` crate.
-> just use the exported function from that crate (you use anyways for the attestation flow).
 
 This also describes better what you are doing:
 that sysfs entry (if it exists) just says if its a SE-guest (not backed by attestation) and can be used as an early exit for e.g. an attestation flow.

File: attestation-agent/kbs_protocol/src/client/rcar_client.rs
Comment (huoqifeng) [reply attestation-agent/kbs_protocol/src/client/rcar_client.rs:191]: If possible, we can have this remain for trouble shooting purpose?

File: attestation-agent/attester/src/se/mod.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs:83]: @steffen-eiden what do you mean? could you please give an example?

File: attestation-agent/attester/src/se/mod.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs:19]: Hopefully, @fitzthum you agree with what @steffen-eiden explained, thank you Steffen!

File: attestation-agent/attester/src/se/mod.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs:75]: Comments refined to explain it.

File: attestation-agent/attester/src/se/mod.rs
Comment (steffen-eiden) [reply attestation-agent/attester/src/se/mod.rs:85]: have a look on [Rust by Example](https://doc.rust-lang.org/rust-by-example/flow_control/match/destructuring/destructure_structures.html):
```
    let faa = Foo { x: (1, 2), y: 3 };

    // You do not need a match block to destructure structs:
    let Foo { x : x0, y: y0 } = faa;
    println!("Outside: x0 = {x0:?}, y0 = {y0}");

```

Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs]: I never used it, easy and powerful, cool!

File: attestation-agent/attester/src/se/mod.rs
Comment (fitzthum) [reply attestation-agent/attester/src/se/mod.rs:19]: You can use the function that @steffen-eiden mentions. Otherwise, fixup the style of this function by using an `if let`. 

> that sysfs entry (if it exists) just says if its a SE-guest (not backed by attestation) and can be used as an early exit for e.g. an attestation flow.

This detection function should only be triggered if attestation is supported. Is that what is happening here?

**Comment by huoqifeng** [comment]
> @huoqifeng I noticed the CI will not alert about code that doesn't compile in the se-attester module. I suppose due to the target-toggle. Do you plan to add CI for s390? If not, maybe we should revisit that toggle, otherwise the code will break unnoticed w/ future refactorings.

I know @BbolroC is working on this, maybe, we bring back the change in lib.rs to enable CI for se?

File: attestation-agent/attester/src/se/mod.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs]: I just noticed @steffen-eiden made it public. Yeah, I'll just use it directly.

File: attestation-agent/attester/src/se/mod.rs
Comment (huoqifeng) [reply attestation-agent/attester/src/se/mod.rs]: Let me remove the unused user_data for now.

**Review by fitzthum — APPROVED** [review]
LGTM

Some kind of test coverage would be good. Maybe that is a follow-on?

**Comment by BbolroC** [comment]
For reviewers, you can find CI enablement for s390x at #576 . Thanks! 😉

**Pull request merged**
